src/main/scala/cromiam/webservice/CromIamApiService.scala

```diff
@@ -44,6 +53,50 @@ trait CromIamApiService extends Directives with SprayJsonSupport with DefaultJso
     Http().singleRequest(cromwellRequest)
   }
 
+  private[webservice] def authorizeThenForwardToCromwell(user: Option[String], workflowId: String, action: String, request: HttpRequest): Future[HttpResponse] = {
+    (for {
+      _ <- completesOnceUploadComplete(request)
+      _ <- requestAuth(WorkflowAuthorizationRequest(user.getOrElse("anon"), workflowId, action))
+      resp <- forwardToCromwell(request)
+    } yield resp) recover {
+      case SamIamDenialException => HttpResponse(status = Unauthorized, entity = "{}")
+    }
+  }
+
+  private[webservice] def authorizeReadThenForwardToCromwell(user: Option[String], workflowId: String, request: HttpRequest): Future[HttpResponse] =
+    authorizeThenForwardToCromwell(user, workflowId, "read", request)
+  private[webservice] def authorizeAbortThenForwardToCromwell(user: Option[String], workflowId: String, request: HttpRequest): Future[HttpResponse] =
+    authorizeThenForwardToCromwell(user, workflowId, "abort", request)
+
+  private[webservice] def authorizeSubmitThenForwardToCromwell(user: Option[String], request: HttpRequest, batch: Boolean): Future[HttpResponse] = {
+
+    def extractWorkflowIds(response: HttpResponse): Future[List[String]] = {
+      if (batch) {
+        Unmarshal(response).to[List[CromwellStatus]].map(_.map(_.id))
+      } else {
+        Unmarshal(response).to[CromwellStatus].map(s => List(s.id))
+      }
+    }
+
+    val userNonOptional = user.getOrElse("anon")
+    (for {
+      _ <- completesOnceUploadComplete(request)
+      _ <- requestAuth(CreateWorkflowAuthorizationRequest(userNonOptional))
+      resp <- forwardToCromwell(request)
+      workflowIds <- extractWorkflowIds(resp)
+      _ <- registerCreation(userNonOptional, workflowIds)
+    } yield resp) recover {
+      case SamIamDenialException => HttpResponse(status = Unauthorized, entity = "{}")
+    }
+  }
+
+  // To prevent "you replied before the upload finished" errors
+  // TODO: Doesn't work. Waiting for lightbend response.
+  private def completesOnceUploadComplete(r: HttpRequest): Future[Unit] = {
```

Comment:
I am curious, what is going on with this function? What is it supposed to do? And probably most importantly, will I run into this issue?